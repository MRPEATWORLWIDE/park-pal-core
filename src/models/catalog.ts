import { NavigationPath } from '#components/Navigation.d'
import { Catalog, DeepFindResult, Taxon } from '#data/catalogs'
import { getSearchUrl } from './url'

export const getRootNavigationLinks = (catalog: Catalog): NavigationPath => {
  return {
    parent: [],
    current: {
      key: 'home',
      href: '/',
      text: 'Home'
    },
    children: catalog.taxonomies[0].taxons.map(taxon => ({
      key: taxon.key,
      href: getSearchUrl(taxon.slug),
      text: taxon.label
    }))
  }
}

export const getNavigationLinks = (taxon: DeepFindResult<Taxon>): NavigationPath => {
  return {
    parent: taxon.memo.map(({ key, slug, label }) => ({
      key: key,
      href: getSearchUrl(slug),
      text: label
    })),
    current: {
      key: taxon.result.key,
      href: getSearchUrl(taxon.result.slug),
      text: taxon.result.label
    },
    children: taxon.result.taxons.map(({ key, slug, label }) => ({
      key: key,
      href: getSearchUrl(slug),
      text: label
    }))
  }
}
