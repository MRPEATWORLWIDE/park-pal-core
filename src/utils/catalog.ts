import type { Catalog, Taxon, Taxonomy } from '#data/catalogs'
import type { DeepFindResult } from '#utils/collection'
import { getSearchUrl } from '#utils/url'
import type { NavigationPath } from '@typings/navigation.d'

const getPrimaryTaxonomy = (catalog: Catalog): Taxonomy => {
  // TODO: taxonomies[0] is a requirement. First taxonomy is considered the navigation one. Should it be configurable?
  return catalog.taxonomies[0]
}

export const getRootNavigationLinks = (catalog: Catalog): NavigationPath => {
  return {
    parent: [],
    current: {
      key: 'home',
      href: '/',
      text: 'Home'
    },
    children: getPrimaryTaxonomy(catalog).taxons.map(taxon => ({
      key: taxon.key,
      href: getSearchUrl(taxon.slug),
      text: taxon.label
    }))
  }
}

export const getNavigationLinks = (taxon: DeepFindResult<Taxon>): NavigationPath => {
  return {
    parent: taxon.memo.map(({ key, slug, label }) => ({
      key: key,
      href: getSearchUrl(slug),
      text: label
    })),
    current: {
      key: taxon.result.key,
      href: getSearchUrl(taxon.result.slug),
      text: taxon.result.label
    },
    children: taxon.result.taxons.map(({ key, slug, label }) => ({
      key: key,
      href: getSearchUrl(slug),
      text: label
    }))
  }
}

export const getSlugs = (catalog: Catalog): string[] => {
  function getFlatSlug(taxon: Taxon): string[] {
    return [taxon.slug].concat(taxon.taxons.flatMap(getFlatSlug))
  }

  return getPrimaryTaxonomy(catalog).taxons.flatMap(getFlatSlug)
}
