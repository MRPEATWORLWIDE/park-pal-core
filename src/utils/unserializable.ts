import { z } from 'zod'

export type Unserializable<T> = {
  _unserializable: Symbol
  data: T
}

export function makeUnserializable<T>(data: T): Unserializable<T> {
  return {
    _unserializable: Symbol.for('unserializable'),
    data
  }
}

export const unserializableSchema = <T extends z.ZodTypeAny>(data: T) => z.object({
  _unserializable: z.unknown().default(Symbol.for('unserializable')),
  data
})