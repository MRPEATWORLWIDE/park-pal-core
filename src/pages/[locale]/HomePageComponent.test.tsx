import type { Catalog, Taxon, Taxonomy } from '#data/catalogs'
import { getRootNavigationLinks } from '#utils/catalog'
import { makeUnserializable } from '#utils/unserializable'
import { render } from '@testing-library/react'
import { I18nProvider } from 'next-localization'
import { HomePageComponent } from './HomePageComponent'

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/',
      pathname: '',
      query: {
        locale: 'en-US'
      },
      asPath: '',
    };
  },
}));

test('home', () => {
  const taxon: Taxon = {
    id: 'taxon_1',
    label: 'Accessories',
    description: 'Accessories',
    slug: 'accessories',
    name: 'All Accessories',
    image: '/assets/images/samsung-case-samsung-galaxy-s21-plus-lifestyle-4-60f836d4aaa9d.png',
    references: [],
    taxons: []
  }

  const taxonomy: Taxonomy = {
    id: 'taxonomy_1',
    facetKey: 'categories',
    name: 'Default Category',
    taxons: [taxon]
  }

  const catalog: Catalog = makeUnserializable({
    id: 'catalog_1',
    name: 'AMER',
    productDataset: {},
    taxonomies: [taxonomy]
  })

  const { container } = render(
    <I18nProvider lngDict={{ general: { viewAll: 'View all' } }} locale='en'>
      <HomePageComponent navigation={getRootNavigationLinks(catalog)} homepage={[]} />
    </I18nProvider>
  )

  // const main = within(screen.getByRole('main'))

  // expect(
  //   main.getByRole('heading', { level: 3, name: /Accessories/i })
  // ).toBeDefined()

  expect(container).toMatchSnapshot()
})
