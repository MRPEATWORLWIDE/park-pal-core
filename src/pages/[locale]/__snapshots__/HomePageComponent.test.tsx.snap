// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home 1`] = `
<div>
  <main>
    <div
      class="container mx-auto px-6"
    >
      <header
        class="border-b-gray-200 border-b pb-6 sticky top-0 bg-white z-50"
      >
        <nav
          class="flex items-center justify-between flex-wrap mb-4"
        >
          <div
            class="flex items-center flex-no-shrink text-white mr-6"
          >
            <a
              href="/en-US"
            >
              <div
                class="py-6 pb-6 "
              >
                <img
                  alt="Commerce Layer logomark"
                  class="inline"
                  src="https://data.commercelayer.app/assets/logos/glyph/black/commercelayer_glyph_black.svg"
                  width="32"
                />
                <img
                  alt="Commerce Layer logotype"
                  class="ml-4 hidden md:inline"
                  src="https://data.commercelayer.app/assets/logos/logotype/black/commercelayer_logotype_black.svg"
                  width="192"
                />
              </div>
            </a>
          </div>
          <div
            class="flex items-center w-auto flex-grow justify-end"
          />
        </nav>
        <div
          class="flex items-center justify-between flex-wrap gap-4 relative"
        >
          <div
            class="order-2 lg:order-1"
          >
            <nav
              class="hidden lg:block lg:relative absolute right-0 top-full w-full z-10 flex-col bg-white py-4 lg:py-0"
            >
              <a
                class="text-back font-bold p-2 m-2"
                href="/en-US/search/accessories"
              >
                Accessories
              </a>
            </nav>
            <button
              aria-controls="menu"
              aria-label="Open menu"
              class="w-11 h-11 flex flex-col items-center justify-center space-y-1 bg-black rounded block lg:hidden"
              type="button"
            >
              <span
                class="w-4 h-0.5 bg-white"
              />
              <span
                class="w-4 h-0.5 bg-white"
              />
              <span
                class="w-4 h-0.5 bg-white"
              />
            </button>
          </div>
          <div
            class="order-1 grow lg:grow-0"
          >
            <label
              class="relative py-2.5 rounded bg-gray-100 text-gray-400 focus-within:text-gray-600 block"
              for="email"
            >
              <svg
                class="pointer-events-none w-6 h-6 absolute top-1/2 transform -translate-y-1/2 left-3"
              />
              <form
                action="/demo-store/en-US/search"
              >
                <input
                  class="form-input appearance-none bg-transparent w-full pl-14 focus:outline-none focus:shadow-outline"
                  placeholder=""
                  value=""
                />
              </form>
            </label>
          </div>
        </div>
      </header>
      <div
        class="mt-10 flex flex-col gap-8"
      />
    </div>
    <div
      class="py-8 lg:p-16 mt-24 bg-gray-50"
    >
      <div
        class="container mx-auto px-6"
      >
        <div
          class="lg:flex items-center justify-between flex-wrap"
        >
          <div>
            <div
              class="flex items-center justify-between"
            >
              <span
                class="flex items-center"
              >
                <svg
                  class="mr-1.5"
                />
                 
                United States
              </span>
              <a
                class="uppercase text-violet-400 text-xs border-b border-gray-200 ml-3 mt-1"
                href="/"
              />
            </div>
          </div>
          <div
            class="mt-6 lg:mt-0"
          >
            <select
              class=" inputSelect"
            >
              <option
                selected=""
                value="en"
              >
                ENG
              </option>
              <option
                value="it"
              >
                ITA
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
